agent/grpc-external/services/resource/watch.go

```diff
@@ -0,0 +1,64 @@
+package resource
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/consul/internal/resource"
+	storage "github.com/hashicorp/consul/internal/storage"
+	pbresource "github.com/hashicorp/consul/proto-public/pbresource"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func watchListOpFrom(eventOp pbresource.WatchEvent_Operation) pbresource.WatchListResponse_Operation {
+	switch eventOp {
+	case pbresource.WatchEvent_OPERATION_UPSERT:
+		return pbresource.WatchListResponse_OPERATION_UPSERT
+	case pbresource.WatchEvent_OPERATION_DELETE:
+		return pbresource.WatchListResponse_OPERATION_DELETE
+	default:
+		panic(fmt.Sprintf("unhandled op %s", eventOp.String()))
+	}
+}
+
+func (s *Server) WatchList(req *pbresource.WatchListRequest, stream pbresource.ResourceService_WatchListServer) error {
+	// check type exists
+	_, ok := s.registry.Resolve(req.Type)
+	if !ok {
+		return status.Error(
+			codes.InvalidArgument,
+			fmt.Sprintf("resource type %s not registered", resource.ToGVK(req.Type)),
+		)
+	}
```

Comment:
As we'll probably end up doing this check in all endpoints, it might be worth adding a helper:

```Go
if _, err := s.resolveType(req.Type); err != nil {
	return err
}
```

```Go
func (s *Server) resolveType(typ *pbresource.Type) (*resource.Registration, error) {
	v, ok := s.registry.Resolve(typ)
	if ok {
		return v, nil
	}
	return nil, status.Errorf(
		codes.InvalidArgument,
		"resource type %s not registered", resource.ToGVK(typ),
	)
}
```

Reply:
Good call! Was waiting for some of the other endpoints to come together but we can do it now. Fixed in https://github.com/hashicorp/consul/pull/16726/commits/7aaffecd6cf378fc95d6797a1ed9afa1f3d6601f